### DISCUSSION TITLE
Useful errors shouldn’t be redacted for non-admin users

### DISCUSSION THREAD

**Discussion by LeBenLeBen** [directus/directus#17035] [Feature Requests]
Currently all error messages returned by the GraphQL API are redacted when you’re not an admin. While I understand the security reasons behind this choice, this has an important impact on the user experience when errors relates to validation & permissions.

For example using the `auth_login` mutation, entering a wrong email/password pair, the following payload is returned (HTTP code 200):

```json
{
	"errors": [
		{
			"message": "An unexpected error occurred.",
			"extensions": {
				"code": "INTERNAL_SERVER_ERROR"
			}
		}
	],
	"data": {
		"auth_login": null
	}
}
```

1. As a developer, I have no way to identify this error as related to the user input and therefor cannot act accordingly
2. As a user, I think that the server has issues instead of knowing that my credentials are wrong

Wouldn’t it be safe enough to return the error message and redact only the stack trace and other technical details from the error? Would you be open as I submit a PR with such a change?

Or is there another way to properly handle such cases and display a proper error to the user that I overlooked?

Related code: https://github.com/directus/directus/blob/0ac08da7ed9d5e4cef52d055be5259fc87ff23ae/api/src/services/graphql/utils/process-error.ts

**Comment by rijkvanzanten** [comment]
I'm wondering if this is a bug or not. The `INTERNAL_SERVER_ERROR` is a catch-all 500 for when the actual error isn't know. A failed login should throw a more specific error regardless (the http 200 status is correct though, that's per the graphql spec). Mind opening a new issue around the wrong error returned? Wondering if it's swallowing things too much

**Reply by LeBenLeBen** [reply]
It’s a global issue, it is not specific to one error, it gets swallowed on purpose [here](https://github.com/directus/directus/blob/0ac08da7ed9d5e4cef52d055be5259fc87ff23ae/api/src/services/graphql/utils/process-error.ts#L16-L21).

Changing these lines for:

```
		return {
			message: error.message,
			extensions: {
				code: 'INTERNAL_SERVER_ERROR',
			},
		};
```

Would be enough to display the actual error on the front-end side.

**Reply by rijkvanzanten** [reply]
That'll open up a security vulnerability, as we don't know what the message is in error.message, as this also handles unknown errors

**Reply by Romakita** [reply]
Hi guys,

I just encountered an issue with that. If the cms is consumed with an account that not have admin role, the graphQL client doesn't have the exact message error. But it useful to have it when it's a graphQL validation error.
For example, with admin mode I have this result:

```json
{
  "errors": [
    {
      "message": "Variable \"$data\" got invalid value \"en-US\" at \"data[0].translations[0].languages_code\"; Expected type \"create_languages_input\" to be an object.",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      },
      "locations": [
        {
          "line": 1,
          "column": 26
        }
      ]
    }
]
```

I think, have this detailed message without admin role is really important for the Developer experience. The real security issue is around the stack trace @rijkvanzanten and not around the error message itself.

I propose to fix the current code by something like that:

```js
const processError = (accountability: Accountability | null, error: Readonly<GraphQLError>): GraphQLFormattedError => {
	logger.error(error);

       // solution1: maybe qualifying error like error validation should be handled here to return the appropriate formattedError 
       // if (error instanceof ValidationError) { 
       //   return {
       //    message: error.message,
       //    extensions: {
       //.      code: 'VALIDATION_ERROR'
       //    }
       //  }
       //}

	if (accountability?.admin === true) {
		return {
			...error,
			extensions: {
				code: 'INTERNAL_SERVER_ERROR',
			},
		};
	} else {
               // solution 2: allow to return the message if the message exists
		return {
			message: error.message || 'An unexpected error occurred.',
			extensions: {
				code: 'INTERNAL_SERVER_ERROR',
			},
		};
	}
};

```
I just want to know if this solution can be a potential acceptable fix for you ? 
See you
Romain

**Reply by rijkvanzanten** [reply]
> The real security issue is around the stack trace @rijkvanzanten and not around the error message itself.

Agree to disagree. For example if you're reading a single item; if the error is a "not found" instead of a "forbidden", it is now leaking what primary keys exist in the system. Similarly, returning the validation errors verbatim exposes the shape of the data model. The only solution here is to make it configurable either for the whole system or per role

**Reply by Romakita** [reply]
I understand your point, it could be a security issue if the graqphql is exposed publicly without bearer auth. I forgot to give you the context. In our case the graphql required a bearer token and a specific role. But currently, to have a proper error validation message is to give a admin role. 
your given solution is clearly that we need and think it’ll be interesting for any project.
I think an optin per role is enough, activating it for the whole system could be a risk if the cms is configured to auto assign a role (or anonymous role).

how I can help you? Is it something simple to implement it?

see you

**Reply by Romakita** [reply]
@rijkvanzanten just another point when you said « **exposes the shape of the data model.** »:
By nature graphql exposes the shape of data model. 
So having a validation error that highlight a problem from a data sent with a gql query doesn't seem like a risk to me (not more that exposing the schema itself).

But maybe I'm wrong. I have the knowledge around the different types of validation errors that could occur on directus. Are these errors are emitted by graphql or are these errors that could come from the database (in which case it’s problematic). 
I’m just curious ;)

see you

**Reply by rijkvanzanten** [reply]
> By nature graphql exposes the shape of data model.

That's not guaranteed! By default introspection will only return the schema information about the fields you have explicit permission to whereas validation can fail on fields you can't access directly, and introspection can be disabled altogether through an environment variable 🙂 

> But maybe I'm wrong. I have the knowledge around the different types of validation errors that could occur on directus. Are these errors are emitted by graphql or are these errors that could come from the database (in which case it’s problematic).
I’m just curious ;)

Both! It could be graphql, the validation layer in between, role permissions, a hook, a flow, or a database error

**Reply by Romakita** [reply]
I see. Thanks a lot for the explanation :)

From what I understand, we should already limit the types of errors that can be reported when the role has the right rights

For an mvp, having GraphQL validation errors would be sufficient if the following conditions are met:

- Have introspection enabled
- Have the appropriate right for the role
- that the error type is GRAPHQL_VALIDATION

While doing some tests, I noticed that when it is a query, we have GRAPHQL_VALIDATION error, but it seems to be less simple in the case of a mutation. Here is my test:

```gql
query getSupportedLanguages {
  languages {
    test // invalid field
    code 
    name
  }
}
```

Gives:

```json
{
  "errors": [
    {
      "message": "GraphQL validation error.",
      "extensions": {
        "code": "GRAPHQL_VALIDATION",
        "errors": [
          {
            "message": "Cannot query field \"test\" on type \"languages\".",
            "locations": [
              {
                "line": 3,
                "column": 5
              }
            ]
          }
        ]
      }
    }
  ]
}
```

When we submitting a mutation, it seems to be more tricky, because the validation error isn't typed:

```gql
mutation createSpaCareCategories ($data: [create_spa_care_sections_input!]!) {
  create_spa_care_sections_items(data: $data) {
    id
    text_id
    uuid
    translations {
      languages_code {
        code
      }
      title
    }
  }
}
```
Variables:
```json
{
  "data": [{
    "test": "1"
  }]
}
```

Gives:

```json
{
  "errors": [
    {
      "message": "Variable \"$data\" got invalid value { test: \"1\" } at \"data[0]\"; Field \"test\" is not defined by type \"create_spa_care_sections_input\".",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      },
      "locations": [
        {
          "line": 1,
          "column": 35
        }
      ]
    }
  ]
}
```
> But it was on the transfer requests that I really needed to have more details.

Same situation with this kind of mutation
```json
{
  "data": [{
    "uuid": 1
  }]
}
```
Gives:

```json
{
  "errors": [
    {
      "message": "Variable \"$data\" got invalid value 1 at \"data[0].uuid\"; String cannot represent a non string value: 1",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      },
      "locations": [
        {
          "line": 1,
          "column": 35
        }
      ]
    }
  ]
}
```

Obviously in the case of mutations the messages are very detailed and relevant but do not seem to be systematically emitted by GraphQL. As there is no distinction, it is more difficult to establish a fine rule of display according to the role and permissions.

But given the format of the message, I believe it comes from graphQL (maybe I'm wrong).

See you

**Reply by rijkvanzanten** [reply]
Oh interesting! Those last two examples look like GraphQL errors that were erroneously seen as an unexpected error 🤔 That might actually be more of a bug than the intended behavior. To your point, if GraphQL is the one throwing the validation error, you already know the schema involved in the request so it should just be resolved as is.

**Reply by rijkvanzanten** [reply]
https://github.com/directus/directus/issues/21264!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
